neo/Consensus/ConsensusService.cs

```diff
@@ -252,49 +293,64 @@ private void OnPrepareRequestReceived(ConsensusPayload payload, PrepareRequest m
 
         private void OnPrepareResponseReceived(ConsensusPayload payload, PrepareResponse message)
         {
-            if (context.Signatures[payload.ValidatorIndex] != null) return;
+            if (context.Preparations[payload.ValidatorIndex] != null) return;
```

Comment:
Can someone remind me where the code verifies that a malicious Validator does not send a `ValidatorIndex` from another validator rather than his own?

Reply:
Please disregard, I figured it out how it is verified. `GetScriptHashesForVerifying` in `ConsensusPayload` uses the `ValidatorIndex` from the payload to verify during normal verification of the message... nice design.

Reply:
Exactly.